src/main/msp/msp.c

```diff
@@ -2318,20 +2381,121 @@ static mspResult_e mspProcessInCommand(uint8_t cmdMSP, sbuf_t *src)
+                vtxSettingsConfigMutable()->freq = (vtxSettingsConfigMutable()->band == 0) ? newFreq : 0;
+            }
+        }
+        break;
+#endif
+
+#ifdef USE_VTX_TABLE
+    case MSP_SET_VTXTABLE_BAND:
+        {
+            char bandName[VTX_TABLE_BAND_NAME_LENGTH + 1];
+            memset(bandName, 0, VTX_TABLE_BAND_NAME_LENGTH + 1);
+            uint16_t frequencies[VTX_TABLE_MAX_CHANNELS];
+            const uint8_t band = sbufReadU8(src);
+            const uint8_t bandNameLength = sbufReadU8(src);
+            for (int i = 0; i < bandNameLength; i++) {
+                const char nameChar = sbufReadU8(src);
+                if (i < VTX_TABLE_BAND_NAME_LENGTH) {
+                    bandName[i] = toupper(nameChar);
+                }
+            }
+            const char bandLetter = toupper(sbufReadU8(src));
+            const bool isFactoryBand = (bool)sbufReadU8(src);
+            const uint8_t channelCount = sbufReadU8(src);
+            for (int i = 0; i < channelCount; i++) {
+                const uint16_t frequency = sbufReadU16(src);
+                if (i < vtxTableConfig()->channels) {
+                    frequencies[i] = frequency;
                 }
             }
+
+            if (band > 0 && band <= vtxTableConfig()->bands) {
+                vtxTableStrncpyWithPad(vtxTableConfigMutable()->bandNames[band - 1], bandName, VTX_TABLE_BAND_NAME_LENGTH);
+                vtxTableConfigMutable()->bandLetters[band - 1] = bandLetter;
+                vtxTableConfigMutable()->isFactoryBand[band - 1] = isFactoryBand;
+                for (int i = 0; i < vtxTableConfig()->channels; i++) {
+                    vtxTableConfigMutable()->frequency[band - 1][i] = frequencies[i];
+                }
+                vtxTableInit();  // Reinitialize and refresh the in-memory copies
+            } else {
+                return MSP_RESULT_ERROR;
+            }
+        }
+        break;
+
+    case MSP_SET_VTXTABLE_POWERLEVEL:
+        {
+            char powerLevelLabel[VTX_TABLE_POWER_LABEL_LENGTH + 1];
+            memset(powerLevelLabel, 0, VTX_TABLE_POWER_LABEL_LENGTH + 1);
+            const uint8_t powerLevel = sbufReadU8(src);
+            const uint16_t powerValue = sbufReadU16(src);
+            const uint8_t powerLevelLabelLength = sbufReadU8(src);
+            for (int i = 0; i < powerLevelLabelLength; i++) {
+                const char labelChar = sbufReadU8(src);
+                if (i < VTX_TABLE_POWER_LABEL_LENGTH) {
+                    powerLevelLabel[i] = toupper(labelChar);
+                }
+            }
+
+            if (powerLevel > 0 && powerLevel <= vtxTableConfig()->powerLevels) {
+                vtxTableConfigMutable()->powerValues[powerLevel - 1] = powerValue;
+                vtxTableStrncpyWithPad(vtxTableConfigMutable()->powerLabels[powerLevel - 1], powerLevelLabel, VTX_TABLE_POWER_LABEL_LENGTH);
+                vtxTableInit();  // Reinitialize and refresh the in-memory copies
```

Comment:
Wondering if it wouldn't be better if we called `vtxTableInit()` only once, before saving the settings?

Reply:
The problem is that the bands and powerlevel MSP messages are iterative and async. We don't know how many are coming from the client. We would hope that it would equal the configured counts but can't guarantee it. Otherwise to only do it once before save we'd need to add another message or embed it in the generalized config saving process (which has no context to the vtxtable updates). Also `vtxTableInit()` is only populating the shadow in-memory copies of the data so it doesn't hurt anything to call repeatedly.

Reply:
It could be done in a way that any call to a VTX table update MSP command set a `vtxTablesNeedInit` boolean, and `MSP_EEPROM_WRITE` checked this, and if required called `vtxTableInit()`. We are already doing something similar for changes to features.

I am aware that `vtxTableInit()` is reentrant at the moment, but I am concerned that this requirement, and the fact that it is called on an incomplete configuration, are non-obvious, and that future modifications to it (like validating the settings for `vtxSettingsConfig()->band` and `vtxSettingsConfig()->channel` against what is defined in the VTX tables) might make this behaviour problematic.
So if we stay with this, then we should at least add a warning to the definition of `vtxTableInit()`, stating that it is called on incomplete configurations.

Reply:
> It could be done in a way that any call to a VTX table update MSP command set a vtxTablesNeedInit boolean, and MSP_EEPROM_WRITE checked this, and if required called vtxTableInit(). We are already doing something similar for changes to features.

Ok, I'll do it this way.